crates/noirc_evaluator/src/ssa/conditional.rs

```diff
@@ -750,6 +733,92 @@ impl DecisionTree {
         }
     }
 
+    // returns condition1 AND condition2
+    fn and_conditions(
+        condition1: Option<NodeId>,
+        condition2: Option<NodeId>,
+        stack_frame: &mut StackFrame,
+        ctx: &mut SsaContext,
+    ) -> Option<NodeId> {
+        match (condition1, condition2) {
+            (None, None) => None,
+            (Some(condition), None) | (None, Some(condition)) => {
+                if condition.is_dummy() {
+                    None
+                } else {
+                    Some(condition)
+                }
+            }
+            (Some(cond1), Some(cond2)) => {
+                if cond1.is_dummy() {
+                    Self::and_conditions(None, condition2, stack_frame, ctx)
+                } else if cond2.is_dummy() {
+                    Self::and_conditions(None, condition1, stack_frame, ctx)
+                } else if cond1 == cond2 {
+                    condition1
+                } else {
```

Comment:
I think we can clean these cases up a bit. The `cond.is_dummy()` case seems to be repeated for both conds. We can reorganize it to avoid this:
```rs
(Some(cond), other) | (other, Some(cond)) if cond.is_dummy() => Self::and_conditions(None, other, stack_frame, ctx)
(Some(cond), None) | (None, Some(cond)) => Some(cond),
(Some(cond1), Some(cond2)) if cond1 == cond2 => condition1,
(Some(cond1), Some(cond2) => ... the else case ...
```
It does seem odd to me that we have optional values which themselves may be dummy values. It'd be nice if we could comprehensively stick to only optional values and remove the dummy values to only have 1 notion of missing value.

Reply:
Done.
I am not sure if we could get dummy values here, but in case of I prefer to handle all the possible cases.